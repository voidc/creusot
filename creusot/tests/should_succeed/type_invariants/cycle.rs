extern crate creusot_contracts;
use creusot_contracts::{*, invariant::Invariant};

pub struct WithInvariant;

impl Invariant for WithInvariant {
    #[predicate]
    #[creusot::type_invariant]
    fn invariant(self) -> bool {
        self.pred()
    }
}

impl WithInvariant {
    #[predicate]
    fn pred(self) -> bool {
        true
    }
}

pub fn id(x: WithInvariant) -> WithInvariant {
    x
}