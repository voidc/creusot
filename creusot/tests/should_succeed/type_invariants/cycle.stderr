error[creusot]: encountered a cycle during translation: [{DefId(0:12 ~ cycle[9581]::{impl#0}::invariant)}, {DefId(0:13 ~ cycle[9581]::{impl#1}::pred)}, {DefId(0:12 ~ cycle[9581]::{impl#0}::invariant)}]
 --> cycle.rs:9:5
  |
9 |     fn invariant(self) -> bool {
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

